### PR TITLE
feature: Define supernode components and interfaces

### PR DESCRIPTION
**Pull request by Starnop**
Signed-off-by: Starnop <starnop@163.com>

### Ⅰ. Describe what this PR did
After offline discussion, the initial version of components and interfaces of supernode are thrown out for review. We have defined the [API](https://github.com/dragonflyoss/Dragonfly/blob/master/apis/swagger.yml) with swagger before.  

Before we talking about the compoents and interfaces, some nouns in the system of Dragonfly/Supernode need to be declared :
1. **Task**: A Task will store some meta info about the taskFile, pieces and something else.A Task has a one-to-one correspondence with a file on the disk which is identified by taskID.
2. **DfgetTask**: A DfgetTask represents **a download process** initiated by dfget or other clients.
3. **Peer**: Each peer launches **a webserver** that provides downloaded files for other nodes to download in the p2p network.

Based on the API and the function of the system,  we have five modules so far as follows:
1. TaskMgr
2. PeerMgr
3. DfgetTaskMgr
4. CDNMgr
5. SchedulerMgr

Based on the above definitions, we can easily guess that each component is responsible for manipulating the corresponding resource object with component name. However,  we have to have a mention about `SchedulerMgr`, the `SchedulerMgr` has two main responsibilities:
1. Responsible for maintaining the correspondence between peer and pieces.
2. Responsible for calculating scheduling results according to certain rules.

And how do these components work together? At present, we mainly have three workflow, I will show the data flow between them through that.
1. Register

image

Note: The CDN is triggered to download the file from source only when the file not exist.

2. Download

image

Note: This is a loop process.

3. ServiceDown

image

Note: When `Dfget` start to download a file form p2p network, it will start a uploader server for provide service for other peers in the p2p network. However,  when uploader server lasts in five minutes without downloading or uploading task, it will automatically sends a request to tell the Supernode that the server will be offline.

### Ⅱ. Does this pull request fix one issue?

None.

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
None.

### CODE REVIEW

**Comment by zhouhaibing089** [comment]
I'm always curious on whether `CDN` is the right term in dragonfly context. CDN generally means content delivery network, do we really mean that?

File: supernode/daemon/mgr/SchedulerMgr.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+	"github.com/dragonflyoss/Dragonfly/apis/types"
+)
+
+// SchedulerMgr has two main responsibilities:
+// 1. Responsible for maintaining the correspondence between peer and pieces.
+// 2. Responsible for calculating scheduling results according to certain rules.
+type SchedulerMgr interface {
+	// InitProgress init the correlation information between peers and pieces, etc.
+	InitProgress(ctx context.Context, taskID, cid string) error
```
Comment (zhouhaibing089) [inline supernode/daemon/mgr/SchedulerMgr.go]: Could we start avoiding writing `cid` any more, people may start asking: why `taskID` is not written as `tid`. It is `clientID`, right?

File: supernode/daemon/mgr/SchedulerMgr.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+	// UpdateProgress update the correlation information between peers and pieces.
+	// 1. update the info about srcCid to tell the scheduler that corresponding peer has the piece now.
+	// 2. update the info about dstCid to tell the scheduler that someone has downloaded the piece form here.
+	// Scheduler will calculate the load and times of error/success for every peer to make better decisions.
+	UpdateSrcProgress(ctx context.Context, taskID, srcCid, dstCid string, pieceRange string, pieceStatus string) error
```
Comment (zhouhaibing089) [inline supernode/daemon/mgr/SchedulerMgr.go]: I am not sure that pieceRange and pieceStatus can be just strings..

I also doubt whether `InitProgress` function is necessary. Can we do the check here(`UpdateSrcProgress`): if the progress is not tracked previously then initialize it.

File: supernode/daemon/mgr/TaskMgr.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	Register(ctx context.Context, task *types.TaskInfo) (taskID string, err error)
+
+	// Get the task Info with specified taskID.
+	Get(ctx context.Context, taskID string) (*types.TaskInfo, error)
+
+	// List returns the list tasks with filter.
+	List(ctx context.Context, filter map[string]string) ([]*types.TaskInfo, error)
+
+	// CheckTaskStatus check whether the taskID corresponding file exists.
+	CheckTaskStatus(ctx context.Context, taskID string) (bool, error)
```
Comment (zhouhaibing089) [inline supernode/daemon/mgr/TaskMgr.go]: So taskStatus is either `true` or `false`? Also does `TaskInfo` already tell its status some way?

File: supernode/daemon/mgr/DfgetTaskMgr.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+// A DfgetTask represents a download process initiated by dfget or other clients.
+type DfgetTaskMgr interface {
+	// Add a new dfgetTask, we use cid to identify a dfgetTask uniquely.
+	// Cid should be generated by dfget, supernode will use it directly.
+	// NOTE: We should create a new dfgetTask for each download process,
+	//       even if the downloads initiated by the same machine.
+	Add(ctx context.Context, dfgetTask *types.DfGetTask) error
+
+	// Get get a dfgetTask info with specified cid and taskID.
+	Get(ctx context.Context, cid string) (dfgetTask *types.DfGetTask, err error)
```
[the remark this answers is not in the document]
Comment (Starnop) [reply supernode/daemon/mgr/DfgetTaskMgr.go]: I am sorry to make my carelessness, I will delete it right now, THX.

File: supernode/daemon/mgr/SchedulerMgr.go (hunk 1 above)
Comment (Starnop) [reply supernode/daemon/mgr/SchedulerMgr.go]: Yeah, you are right, I will change it.

File: supernode/daemon/mgr/TaskMgr.go (hunk 3 above)
Comment (Starnop) [reply supernode/daemon/mgr/TaskMgr.go]: This is an interface to quickly check if a file exists, and you can use get for more information, including status.

File: supernode/daemon/mgr/PeerMgr.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// PeerMgr as an interface defines all operations against Peer.
+// A Peer represents a web server that provides file downloads for others.
+type PeerMgr interface {
+	// Register a peer with specified peerInfo.
+	// Supernode will generate a unique peerID for every Peer with PeerInfo provided.
+	Register(ctx context.Context, peerInfo *types.PeerInfo) (peerID string, err error)
+
+	// RegistryCDNNode registry a cdn node with specified peerInfo.
+	// NOTE: A CDN Node is equivalent to a supernode actually.
+	RegisterCDNNode(ctx context.Context, peerInfo *types.PeerInfo) error
```
[the remark this answers is not in the document]
Comment (Starnop) [reply supernode/daemon/mgr/PeerMgr.go]: I think about it carefully, and I don't think we need this  interface at present, it is consistent with the implementation of `Register`, as @zhouhaibing089  said, if necessary, we can use a flag to indicate whether it is a CDN node and it's enough. @allencloud @lowzj  WDYT?

File: supernode/daemon/mgr/PeerMgr.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+	// UnRegister offline a peer service and
+	// NOTE: update the info related for scheduler.
+	UnRegister(ctx context.Context, peerID string) error
+
+	// Get the peer Info with specified cid.
+	Get(ctx context.Context, peerID string) (*types.PeerInfo, error)
+
+	// List return a list of peers info with filter.
+	List(ctx context.Context, filter map[string]string) (peerList []*types.PeerInfo, err error)
```
[the remark this answers is not in the document]
Comment (Starnop) [reply supernode/daemon/mgr/PeerMgr.go]: In fact, what type of filter should be used is what I struggle with, maybe we can have more discussion to confirm the final solution. @zhouhaibing089 @lowzj @allencloud

File: supernode/daemon/mgr/SchedulerMgr.go (hunk 2 above)
Comment (Starnop) [reply supernode/daemon/mgr/SchedulerMgr.go]: 1. For pieceRange the format `pieceRange` should be like : `start-end`.
For `PieceStatus`, using strings is so flexible, but if we define a corresponding structure for every status such as `CDNStatus, PieceStatus, etc` is redundant. Any good ideas?
2. I think `InitProgress` and `UpdateProgress` is all needed,  they do the different things separately. Otherwise, `UpdateProgress` have to do so much things.

File: supernode/daemon/mgr/CDNMgr.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package mgr
```
[the remark this answers is not in the document]
Comment (zhouhaibing089) [reply supernode/daemon/mgr/CDNMgr.go]: I mean all lowercased actually, maybe `cdn_manager.go`?

File: apis/types/df_get_task.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
+// Code generated by go-swagger; DO NOT EDIT.
+
+package types
+
+// This file was generated by the swagger tool.
+// Editing this file might prove futile when you re-run the swagger generate command
```
Comment (fengzixu) [inline apis/types/df_get_task.go]: Lacking of the `,`.  Please notice that the comment is the full statement.

File: apis/types/df_get_task.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+import (
+	"encoding/json"
+
+	"github.com/go-openapi/errors"
+	strfmt "github.com/go-openapi/strfmt"
```
Comment (fengzixu) [inline apis/types/df_get_task.go]: Why did you rename this package?

File: apis/types/df_get_task.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/go-openapi/swag"
+	"github.com/go-openapi/validate"
+)
+
+// DfGetTask A download process initiated by dfget or other clients.
```
Comment (fengzixu) [inline apis/types/df_get_task.go]: DfGetTask: A download process initiated by dfget or other clients.

File: apis/types/df_get_task.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+//
+// swagger:model DfGetTask
+type DfGetTask struct {
+
```
Comment (fengzixu) [inline apis/types/df_get_task.go]: The blank line is useless.

File: apis/types/df_get_task.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+	// The status of Dfget download process.
+	//
+	// Enum: [WAITING RUNNING FAILED SUCCESS]
+	Status string `json:"status,omitempty"`
+
+	// task Id
+	TaskID string `json:"taskId,omitempty"`
+}
+
+// Validate validates this df get task
```
Comment (fengzixu) [inline apis/types/df_get_task.go]: Same problem as above.

File: apis/types/df_get_task.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+	DfGetTaskStatusRUNNING string = "RUNNING"
+
+	// DfGetTaskStatusFAILED captures enum value "FAILED"
+	DfGetTaskStatusFAILED string = "FAILED"
+
+	// DfGetTaskStatusSUCCESS captures enum value "SUCCESS"
+	DfGetTaskStatusSUCCESS string = "SUCCESS"
+)
+
+// prop value enum
```
Comment (fengzixu) [inline apis/types/df_get_task.go]: hmmm, the comment looks like nonsense.

File: apis/types/df_get_task.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+func (m *DfGetTask) validateStatusEnum(path, location string, value string) error {
+	if err := validate.Enum(path, location, value, dfGetTaskTypeStatusPropEnum); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *DfGetTask) validateStatus(formats strfmt.Registry) error {
+
```
Comment (fengzixu) [inline apis/types/df_get_task.go]: useless blankline

Comment (Starnop) [reply apis/types/df_get_task.go]: I have to tell you that the code under the folder `apis/types` is automatically generated by swagger. So the code [swagger.yml](https://github.com/dragonflyoss/Dragonfly/blob/master/apis/swagger.yml) should be reviewed instead of under the folder `apis/types`.

Comment (fengzixu) [reply apis/types/df_get_task.go]: I see.

Comment (fengzixu) [reply apis/types/df_get_task.go]: I have deleted the comments above.

**Comment by fengzixu** [comment]
In the https://github.com/dragonflyoss/Dragonfly/pull/284/files#diff-c29f432f2663f7d031c5080465b377c1R536, I found the world `recganises`. Is there a mistake of spelling?
